refactor(connlib): remove `async` from the `Device` API

thomaseizinger opened the pull request:
At present, the definition of `Device` is heavily nested with conditional code. I've found this hard to understand and navigate. Recent refactorings now made it possible to remove a lot of these layers so we primarily deal with two concepts:

- A `Device` which offers async read and non-blocking write functions
- A `Tun` abstraction which is platform-specific

Instead of dedicated modules, I chose to feature-flag individual functions on `Device` with `#[cfg(target_family = "unix")]` and `#[cfg(target_family = "windows")]`. I find this easier to understand because the code is right next to each other.

In addition, changing the module hierarchy of `Device` allows us to remove `async` from the public API which is only introduced by the use of `rtnetlink` in Linux. Instead of making functions across all `Tun` implementations `async`, we embed a "worker" within the `linux::Tun` implementation that gets polled before `poll_read`.

thomaseizinger committed:
Reduce duplication of getting MTU between flavours of UNIX

jamilbk reviewed (commented):
Wow, thanks for the cleanup. I remember hitting a lot of issues around which syscalls are allowed on locked-down platforms like android and darwin -- just for clarity, this PR *does not* change *which* actual syscalls are being made, it just refactors them correct?

thomaseizinger committed:
Create ctor for `Device`

thomaseizinger committed:
Inline `IfaceConfig` into `Device`

thomaseizinger committed:
Inline `create_iface`

thomaseizinger committed:
Move `ioctl` up to `device_channel`

thomaseizinger committed:
Inline `device_channel_unix` and `device_channel_win`

thomaseizinger committed:
Move `AsyncFd` into individual TUN implementations

thomaseizinger committed:
Make `set_interface_config` a free function

thomaseizinger committed:
Handle `up` as part of `new`

thomaseizinger committed:
Introduce `worker` within linux TUN implementation

This allows us to make all functions non-async to great benefit
of the rest of the codebase.

thomaseizinger committed:
Fill in dummy values for windows

thomaseizinger committed:
Fix warnings on Windows

thomaseizinger committed:
Reduce duplication between flavours of unix for polling AsyncFd

thomaseizinger commented on rust/connlib/tunnel/src/device_channel/tun_linux.rs:
```diff
@@ -0,0 +1,267 @@
+    connection: tokio::task::JoinHandle<()>,
+    fd: AsyncFd<RawFd>,
+
+    worker: Mutex<Option<BoxFuture<'static, Result<()>>>>,
```
This `Mutex` will go away with https://github.com/firezone/firezone/pull/2659 because we won't be sharing the `Device` in an `Arc` anymore and thus can just access this `Future` with an `&mut`.

thomaseizinger committed:
Fix Android compile error

conectado reviewed (commented):
Looks good! can merge after we test performance! :)

conectado reviewed (approved):
Tested this both from a vultr instance to aws and from my local computer to aws in a bunch of different setups and saw no performance degradation at all :) 🚀

conectado committed:
Merge branch 'main' into refactor/device-no-async
